File: 0114) Flatten Binary Tree to Linked List.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public void flatten(TreeNode root) {
        if(root != null)
        {
            helper(root);
        }
    }
    public TreeNode head = null;
    public void helper(TreeNode root)
    {
        if(root.right != null)
        {
            helper(root.right);
        }
        if(root.left != null)
        {
            helper(root.left);
        }
        root.left = null;
        root.right = head;
        head = root;
    }
}

File: 0114) Flatten Binary Tree to Linked List.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    void flatten(TreeNode* root) {
        if(root != nullptr)
        {
            convert(root);
        }
    }
    TreeNode *head = nullptr;
    void convert(TreeNode* root)
    {
        if(root->right)
        {
            convert(root->right);
        }
        if(root->left)
        {
            convert(root->left);
        }
        root->left = nullptr;
        root->right = head;
        head = root;
    }
};
